dht_net.go

```diff
@@ -80,6 +81,11 @@ func (dht *IpfsDHT) handleNewMessage(s network.Stream) bool {
 	defer timer.Stop()
 
 	for {
+		if dht.getMode() != ModeServer {
```

Comment:
Not sure, this was grabbed from an existing PR. I think we may be defending from some asynchronicity issues (e.g. we start `handleNewMessage` just as we switch from Server to Client mode)

Reply:
We have a race:

1. We've received and negotiated an inbound DHT stream but haven't yet set the protocol name on the stream itself.
2. We switch to DHT client mode.
3. We unregister the DHT stream handler.
4. We kill all inbound DHT streams. However, we'll miss the stream from 1 because we don't know that it's a DHT stream.
5. We start handling the stream and never stop.

Reply:
Thanks for the explanation, that makes sense. This is why I've tried to leave in as much of the code from the PRs I was migrating as possible, although sometimes it means I don't the exact reasoning.